src/datomish/clauses.cljc

```diff
@@ -0,0 +1,256 @@
+;; This Source Code Form is subject to the terms of the Mozilla Public
+;; License, v. 2.0. If a copy of the MPL was not distributed with this
+;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+(ns datomish.clauses
+  (:require
+     [datomish.source
+      :refer [attribute-in-source
+              constant-in-source
+              source->from
+              source->constraints]]
+     [datomish.util :as util #?(:cljs :refer-macros :clj :refer) [raise-str cond-let]]
+     [datascript.parser :as dp
+      #?@(:cljs
+            [:refer
+             [PlainSymbol Predicate Not Pattern DefaultSrc Variable Constant Placeholder]])]
+     [honeysql.core :as sql]
+     [clojure.string :as str]
+     )
+  #?(:clj
+       (:import
+          [datascript.parser
+           PlainSymbol Predicate Not Pattern DefaultSrc Variable Constant Placeholder])))
+
+;; A CC is a collection of clauses that are combined with JOIN.
+;; The topmost form in a query is a ConjoiningClauses.
+;;
+;; Ordinary pattern clauses turn into FROM parts and WHERE parts using :=.
+;; Predicate clauses turn into the same, but with other functions.
+;; Function clauses with bindings turn into:
+;;   * Subqueries. Perhaps less efficient? Certainly clearer.
+;;   * Projection expressions, if only used for output.
+;;   * Inline expressions?
+;; `not` turns into NOT EXISTS with WHERE clauses inside the subquery to
+;; bind it to the outer variables, or adds simple WHERE clauses to the outer
+;; clause.
+;; `not-join` is similar, but with explicit binding.
+;; `or` turns into a collection of UNIONs inside a subquery.
+;; `or`'s documentation states that all clauses must include the same vars,
+;; but that's an over-simplification: all clauses must refer to the external
+;; unification vars.
+;; The entire UNION-set is JOINed to any surrounding expressions per the `rule-vars`
+;; clause, or the intersection of the vars in the two sides of the JOIN.
+
+;; `from` is a list of [source alias] pairs, suitable for passing to honeysql.
+;; `bindings` is a map from var to qualified columns.
+;; `wheres` is a list of fragments that can be joined by `:and`.
+(defrecord ConjoiningClauses [source from external-bindings bindings wheres])
```

Comment:
`defrecord` doesn't support docstrings, because it generates a class. Putting it in a docstring would involve `alter-meta!`, which is a step too far for me.